lib/KettleServer.js

```diff
@@ -0,0 +1,303 @@
+            handler.gradeNames = [];
+        }
+    }
+    return match;
+};
+
+kettle.server.checkCreateRequest = function (server, req, res, next, originOptions) {
+    var match = kettle.server.evaluateRoute(server, req, originOptions);
+    if (match) {
+        fluid.extend(req, match.output); // TODO: allow match to output to other locations
+        var handler = match.handler;
+        if (handler.prefix) {
+            if (req.url.indexOf(handler.prefix) !== 0) {
+                fluid.fail("Failure in route matcher - request url " + req.url + " does not start with prefix " + handler.prefix + " even though it has been matched");
+            } else {
+                // This is apparently the time-honoured behaviour implemented by all express "routers" - the original url is preserved within
+                // req.originalUrl
+                req.url = req.url.substring(handler.prefix.length);
+                req.baseUrl = handler.prefix;
+            }
+        }
+        var options = fluid.extend({gradeNames: handler.gradeNames}, originOptions);
+        handler.app.requests.events.createRequest.fire({
+            type: handler.type,
+            options: options
+        }, req, res, next);
+    }
+    return req.fluidRequest; // we've either created one or we haven't
+};
+
+kettle.server.getRouter = function (that /*, req, handler */) {
+    return that.httpRouter; // default policy simply returns the single httpRouter
+};
+
+kettle.server.sequenceRequest = function (fullSequence, request) {
+    var sequence = fluid.promise.sequence(fullSequence, request);
+    var togo = fluid.promise();
+    sequence.then(function () { // only the handler's promise return counts for success resolution
+        fluid.promise.follow(request.handlerPromise, togo);
+    }, togo.reject);
+    return togo;
+};
+
+kettle.server.getDispatcher = function (that, rootSequence) {
+    return function (req, res, next, options) {
+        var request = kettle.server.checkCreateRequest(that, req, res, next, options);
+        if (request) {
+            fluid.log("Kettle server allocated request object with type ", request.typeName);
+            var requestSequence = kettle.middleware.getHandlerSequence(request, "requestMiddleware");
+            var fullSequence = rootSequence.concat(requestSequence).concat([kettle.request.handleRequestTask]);
+            var handleRequestPromise = kettle.server.sequenceRequest(fullSequence, request);
+            request.handleFullRequest(request, handleRequestPromise, next);
+            handleRequestPromise.then(kettle.request.clear, kettle.request.clear);
+        } else {
+            fluid.log("Kettle server getDispatcher found no matching handlers, continuing");
+            next();
+        }
+    };
+};
+
+kettle.server.registerOneHandler = function (that, app, handler) {
+    if (!handler) { // A typical style of overriding handlers sets them to `null` in derived grades - ignore these
+        return;
```

Comment:
Thanks for the explanation.  Can we add a TODO here indicating work that should be done once FLUID-4982 is addressed?